### PR TITLE
Model methods that extend Model Manager functionality are now final

### PR DESCRIPTION
**Pull request by SidRoberts**
Hello!

* Type: bug fix

* Link to issue:

**In raising this pull request, I confirm the following:**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/cphalcon/blob/master/CONTRIBUTING.md)
- [x] I have checked that another pull request for this purpose does not exist
- [x] I wrote some tests for this PR
- [-] I updated the CHANGELOG

`Mvc\Model::getSource()` is now final and should be replaced by using `setSource()` in `initialize()`:

```php
class Post extends \Phalcon\Mvc\Model
{
    public function initialize()
    {
        $this->setSource("custom_posts_source");
    }
}
```

Overriding `getSource()` meant that `Model::getSource()` and `Model\Manager::getModelSource()` would return different things.

### CODE REVIEW

**Comment by SidRoberts** [comment]
Will wait until https://github.com/phalcon/cphalcon/pull/13951 is merged and then will rebase and remove the revert commit.

**Comment by niden** [comment]
@SidRoberts 13951 is merged.

**Review by Jeckerson — COMMENTED** [review]
Can you change inside all `$this->setSource();` from double quotes `"` into singles `'` ?

File: tests/integration/Mvc/Model/Refactor-ManagerCest.php [hunk 1]
```diff
@@ -64,6 +65,29 @@ public function testShouldReturnSourceWithPrefix(IntegrationTester $I)
...
+     * @since  2019-04-06
+     */
+    public function testModelSourceShouldBeTheSame(IntegrationTester $I)
+    {
+        $manager = new Manager();
+
+        $robots = new People(null, null, $manager);
+
+        $I->assertEquals(
+            "personas",
```
Comment (Jeckerson) [inline tests/integration/Mvc/Model/Refactor-ManagerCest.php]:
```suggestion
            'personas',
```

File: tests/integration/Mvc/Model/Refactor-ManagerCest.php [hunk 2]
```diff
@@ -64,6 +65,29 @@ public function testShouldReturnSourceWithPrefix(IntegrationTester $I)
...
+            $robots->getSource()
+        );
+
+        $I->assertEquals(
+            "personas",
```
Comment (Jeckerson) [inline tests/integration/Mvc/Model/Refactor-ManagerCest.php]:
```suggestion
            'personas',
```

**Review by sergeyklay — DISMISSED** [review]

**Comment by david-duncan** [comment]
I think this MR exposes the amount of pain upgrading is going to take. 51 files changed for something that has been canon for using Phalcon's ORM for years now.

We may want to look into codemods for changes like this.

**Comment by niden** [comment]
@david-duncan I was too a bit concerned but then looked at the files changed and in reality it was only one file (model.zep) that had the significant work. The rest of the files were the models defined in the testing suite.

Since we marked this as `Breaks BC` users should be aware that this will require additional changes. In reality for an existing application all one will have to do is use `setSource` instead of `getSource`. For instance in my models I use this

```php
function getSource(): string
{
    return 'robots';
}
```
so I will have to remove that and add this in the initialize
```php
$this->setSource('robots');
```

Now I agree this is a change, but it will be more of a time consuming change that will have to happen once vs something more serious. After all, once models are set, they don't change that much, other than adding a bit more functionality here and there.

**Review by niden — APPROVED** [review]

**Comment by niden** [comment]
Thank you @SidRoberts

**Comment by niden** [comment]
@CameronHall @Jeckerson Can we add an issue to the devtools so that this becomes part of the template generating models? Thanks

**Comment by CameronHall** [comment]
Done @niden https://github.com/phalcon/phalcon-devtools/issues/1297

**Pull request merged**